### PR TITLE
Darcy.rayner/add log writer

### PR DESCRIPTION
**Pull request by DarcyRaynerDD**
### What does this PR do?
Adds an option to export traces via logs, as opposed to using the agent.

Traces are outputted as logs in the following format, (except without whitespace):
```json
{
  "datadog_traces": [
    {
      "trace_id": "123456789",
      "span_id": "7312976640099218183",
      "parent_id": "164575748657836115",
      "name": "calculation-long-number",
      "resource": "calculation-long-number",
      "error": 0,
      "meta": {
        "language": "javascript"
      },
      "metrics": {
        "_sample_rate": 1,
        "_sampling_priority_v1": 2
      },
      "start": 1565381107232644000,
      "duration": 260664063,
      "service": "node"
    },
    {
      "trace_id": "123456789",
      "span_id": "164575748657836115",
      "parent_id": "13282148538394768476",
      "name": "aws.lambda",
      "resource": "aws.lambda",
      "error": 0,
      "meta": {
        "language": "javascript"
      },
      "metrics": {
        "_sample_rate": 1,
        "_sampling_priority_v1": 2
      },
      "start": 1565381107070475300,
      "duration": 539684570,
      "service": "node"
    }
  ]
}
```

Also refactors Recorder, into AgentExporter and LogExporter classes, and introduces the SpanProcessor for handling sampling/erasure

### Motivation
Preliminary support for lambda tracing

### Plugin Checklist

- [X] Unit tests.
- [X] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/src/plugins/index.js

### CODE REVIEW

File: packages/dd-trace/src/opentracing/tracer.js [hunk 1]
```diff
@@ -37,9 +38,14 @@ class DatadogTracer extends Tracer {
     this._logInjection = config.logInjection
     this._analytics = config.analytics
     this._prioritySampler = new PrioritySampler(config.env)
-    this._writer = new Writer(this._prioritySampler, config.url)
-    this._recorder = new Recorder(this._writer, config.flushInterval)
-    this._recorder.init()
+
+    if (config.experimental !== undefined && config.experimental.useLogWriter) {
```
Comment (rochdev) [inline packages/dd-trace/src/opentracing/tracer.js]: This should be `useLogExporter`. I wonder if the exporter could be configurable instead, something like `exporter: LOG_EXPORTER`.

File: packages/dd-trace/src/agentless/exporter.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  export (span) {
+    const spanContext = span.context()
+    const trace = spanContext._trace
+
+    log.debug(() => `Adding trace to queue: ${JSON.stringify(trace)}`)
+
+    let size = 0
+    let queue = []
+
+    for (const span of trace.finished) {
```
Comment (rochdev) [inline packages/dd-trace/src/agentless/exporter.js]: This should not happen on every export.

File: packages/dd-trace/src/span_processor.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    if (trace.started.length === trace.finished.length) {
+      this._prioritySampler.sample(spanContext)
+
+      if (spanContext._traceFlags.sampled === false) {
+        log.debug(() => `Dropping trace due to user configured filtering: ${trace}`)
+        this._erase(trace)
+        return
+      }
+      this._exporter.export(span)
```
[the remark this answers is not in the document]
Comment (DarcyRaynerDD) [reply packages/dd-trace/src/span_processor.js]: I think this way plays nicely if we want to give the span processor multiple exporters to submit to. That way, the exporters don't clobber each other by trying to both modify the spans, it's left up to the processor to do it.

File: packages/dd-trace/src/agentless/exporter.js (hunk 2 above)
Comment (DarcyRaynerDD) [reply packages/dd-trace/src/agentless/exporter.js]: Isn't the point of the LogExporter to flush whenever it has a finished span?

File: packages/dd-trace/src/span_processor.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (rochdev) [reply packages/dd-trace/src/span_processor.js]: What I meant is that `format()` is called in the exporter instead of in the processor. I wonder if that should be done in the processor instead. This will be important when we implement user-defined hooks into the processor since at that point they should have a formatted span to work with before it reaches the exporters.

File: packages/dd-trace/src/agentless/exporter.js (hunk 2 above)
Comment (rochdev) [reply packages/dd-trace/src/agentless/exporter.js]: If you want to send individual spans as they are finished, why is the loop needed? You should just have to send the span directly. Also, I thought the idea was to batch multiple spans as a single log entry.

File: packages/dd-trace/src/span_processor.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (DarcyRaynerDD) [reply packages/dd-trace/src/span_processor.js]: Okay, that makes sense

File: packages/dd-trace/src/agentless/exporter.js (hunk 2 above)
Comment (DarcyRaynerDD) [reply packages/dd-trace/src/agentless/exporter.js]: So, the processor, (which calls the exporters), won't try and send a span until the trace it's on has all it's spans completed, which was originally logic that was in the Writer. We don't want to have one span per log line, because it's going to cause a lot of invokes of our log forwarder and be expensive for our customers. Maybe the exporters should take a trace as input instead?

File: packages/dd-trace/src/opentracing/tracer.js (hunk 1 above)
Comment (DarcyRaynerDD) [reply packages/dd-trace/src/opentracing/tracer.js]: I've updated this, I also export those constants in the module, and in the d.ts file

File: packages/dd-trace/src/agentless/exporter.js (hunk 2 above)
Comment (DarcyRaynerDD) [reply packages/dd-trace/src/agentless/exporter.js]: So, what I ended up settling on doing, is sending the exporters a list of spans they need to write out. The processor will only send a list of spans to be exported once the trace is complete.

File: index.d.ts [hunk 4]
```diff
@@ -705,125 +705,125 @@ declare namespace plugins {
...
-  interface router extends Integration {}
+  interface router extends Integration { }
 
-    /**
-   * This plugin automatically instruments the
-   * [tedious](https://github.com/tediousjs/tedious/) module.
-   */
-  interface tedious extends Integration {}
+  /**
+ * This plugin automatically instruments the
```
Comment (rochdev) [inline index.d.ts]: Broken formatting.

File: index.d.ts [hunk 5]
```diff
@@ -833,4 +833,7 @@ declare namespace plugins {
  */
 export declare const tracer: Tracer;
 
+export declare const LOG_EXPORTER: 'log-exporter';
```
Comment (rochdev) [inline index.d.ts]: These should not be exported directly from the index, and instead from the `ext` folder similar to other public constants. Unless the idea was to not make them public, in which case the TS definitions are not needed.

File: packages/dd-trace/src/agentless/exporter.js (hunk 2 above)
Comment (rochdev) [reply packages/dd-trace/src/agentless/exporter.js]: This should be done at the exporter level. Basically the processor sends every span to the exporter and each exporter can then have their own logic on how to buffer them. This will be useful if we want to change the strategy in a single exporter, for example to stream spans.

File: .vscode/settings.json [hunk 6]
```diff
@@ -1,3 +1,3 @@
 {
-  "debug.node.autoAttach": "on"
+  "debug.node.autoAttach": "on",
```
Comment (rochdev) [inline .vscode/settings.json]: Invalid JSON.

File: ext/exporters.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
+'use strict'
+const { LOG_EXPORTER, AGENT_EXPORTER } = require('../packages/dd-trace/src/constants')
+
+module.exports = {
+  LOG_EXPORTER: LOG_EXPORTER,
```
Comment (rochdev) [inline ext/exporters.js]: `_EXPORTER` suffix is redundant.

File: index.d.ts [hunk 8]
```diff
@@ -515,69 +520,69 @@ declare namespace plugins {
    * This plugin automatically instruments the
    * [amqp10](https://github.com/noodlefrenzy/node-amqp10) module.
    */
-  interface amqp10 extends Integration {}
+  interface amqp10 extends Integration { }
```
Comment (rochdev) [inline index.d.ts]: Unnecessary change. If formatting should be reviewed, we should add tooling in the project.

File: index.d.ts [hunk 9]
```diff
@@ -236,6 +236,11 @@ export declare interface TracerOptions {
    */
   experimental?: boolean | {
     b3?: boolean
+    /**
+     * Whether to write traces to log output, rather than send to an agent
+     * @default false
+     */
+    exporter?: 'log-exporter' | 'agent-exporter'
```
Comment (rochdev) [inline index.d.ts]: Is it possible to use the values directly from the constants?

File: packages/dd-trace/index.js [hunk 10]
```diff
@@ -2,11 +2,11 @@
 
 const platform = require('./src/platform')
 const node = require('./src/platform/node')
-
 platform.use(node)
 
 const TracerProxy = require('./src/proxy')
 
 module.exports = new TracerProxy()
+
```
Comment (rochdev) [inline packages/dd-trace/index.js]: Unnecessary change.

File: packages/dd-trace/src/config.js [hunk 11]
```diff
@@ -52,7 +53,9 @@ class Config {
     }
     this.runtimeMetrics = String(runtimeMetrics) === 'true'
     this.experimental = {
-      b3: !(!options.experimental || !options.experimental.b3)
+      b3: !(!options.experimental || !options.experimental.b3),
+      exporter: (typeof options.experimental === 'object' && options.experimental.exporter === LOG_EXPORTER)
```
Comment (rochdev) [inline packages/dd-trace/src/config.js]: This could be shortened as the check for the type is unnecessary.

File: packages/dd-trace/src/constants.js [hunk 12]
```diff
@@ -6,5 +6,7 @@ module.exports = {
   ANALYTICS_KEY: '_dd1.sr.eausr',
   ORIGIN_KEY: '_dd.origin',
   HOSTNAME_KEY: '_dd.hostname',
-  REFERENCE_NOOP: 'noop'
+  REFERENCE_NOOP: 'noop',
+  LOG_EXPORTER: 'log-exporter',
```
Comment (rochdev) [inline packages/dd-trace/src/constants.js]: These are not needed. While this is not clear at a glance, this file is only for private constants that are not exposed externally.

File: packages/dd-trace/src/opentracing/tracer.js [hunk 13]
```diff
@@ -5,8 +5,9 @@ const Tracer = opentracing.Tracer
 const Reference = opentracing.Reference
 const Span = require('./span')
 const SpanContext = require('./span_context')
-const Writer = require('../writer')
-const Recorder = require('../recorder')
+const AgentExporter = require('../exporters/agent/exporter')
```
Comment (rochdev) [inline packages/dd-trace/src/opentracing/tracer.js]: Renaming both exporters to `index.js` would allow importing with `../exporters/agent`

File: packages/dd-trace/src/opentracing/tracer.js [hunk 14]
```diff
@@ -37,9 +38,14 @@ class DatadogTracer extends Tracer {
     this._logInjection = config.logInjection
     this._analytics = config.analytics
     this._prioritySampler = new PrioritySampler(config.env)
-    this._writer = new Writer(this._prioritySampler, config.url)
-    this._recorder = new Recorder(this._writer, config.flushInterval)
-    this._recorder.init()
+
+    if (config.experimental !== undefined && config.experimental.exporter === constants.LOG_EXPORTER) {
```
Comment (rochdev) [inline packages/dd-trace/src/opentracing/tracer.js]: Unnecessary check for `undefined`.

Comment (DarcyRaynerDD) [reply packages/dd-trace/src/opentracing/tracer.js]: If I remove the check, it breaks a lot of tests in a bunch of different files which have been initializing tracer with an incomplete config.

Comment (rochdev) [reply packages/dd-trace/src/opentracing/tracer.js]: I would say to fix the tests as they should not force workarounds in code.

Comment (DarcyRaynerDD) [reply packages/dd-trace/src/opentracing/tracer.js]: I just double checked, I was wrong about that test, just going to follow your original suggestion.

File: index.d.ts (hunk 9 above)
Comment (DarcyRaynerDD) [reply index.d.ts]: I tried to import the constants form ext/exporters/.d.ts, and it wasn't playing nicely with the type check.
